### DISCUSSION TITLE
Request specific continuation

### DISCUSSION THREAD

**Discussion by Michal-Szczepaniak** [LuanRT/YouTube.js#385] [Q&A] [answered]
Is there a way, after using "getContinuationData" to add that to request? To just requeset data of specific continuation? Instead of spamming getContinuation?

**Comment by LuanRT** [comment]
Hi @Michal-Szczepaniak!

Most definitely yes but I think I need more context here.

When you say “after using "getContinuationData" to add that to request”, do you mean getting the next continuation node from the response of that method and then retrieving its data instead of using `getContinuation()`?

If you can, please feel free to provide code/pseudocode that demonstrates what you mean.

**Comment by Michal-Szczepaniak** [comment]
Previously i was using yt-comment-scraper for example and here's simple code
```
import ytcm from "@freetube/yt-comment-scraper";

let query = process.argv[2];
let continuation = process.argv[3];

const payload = {
  videoId: query,
  sortByNewest: false,
  continuation: continuation,
}

ytcm.getComments(payload).then(d => console.log(JSON.stringify(d, null, 2)), e => console.error(JSON.stringify(e, null, 2)));
```

i was printing the result and handling that in my app, then i could just get the continuation data, pass it to the next execution and it would fetch that exact continuation. 

Usage example:
I get channel videos/comments/whatever with continuation. i display that in my app. Now i want to fetch continuation of it. With the data I already have, i should be able to just make one request and get the data from continuation. Instead the only option i see is to get first page again and spam `.getContinuation()`

**Comment by LuanRT** [comment, marked as answer]
Ah, got it! Thanks for clarifying.

Please correct me if I'm wrong here:

* You need the continuation tokens so you can retrieve them later.
* You want to request that continuation without having to make multiple requests.

I think you should be able to do that with the `Actions#execute` method. Here's a small example for comments:

```ts
import { Innertube, UniversalCache, YTNodes } from 'youtubei.js';

(async () => {
  const yt = await Innertube.create({ cache: new UniversalCache(false), generate_session_locally: true });

  const search = await yt.search('Mac Miller right');
  const comments = await yt.getComments(search.videos[0].id);

  // Get the continuation node from the first response. This contains the token and other parameters needed to get the next page.
  const continuation_node = comments.page.on_response_received_endpoints.at(1).contents.firstOfType(YTNodes.ContinuationItem);

  /**
   * The Actions#execute(endpoint, args?) method takes an endpoint and payload and then proceeds to make the request.
   * I recommend that you save the entire continuation node in your app so you don't have to manually 
   * create the payload and find the correct api url. In case your app can only save the token, you
   * may use the "/next" endpoint and pass the token as "continuation" or "token" in the args.
   * 
   * See the example below. Which uses the metadata and endpoint from the continuation node
   * to make the request with all the parameters needed. 
   * 
   * Note that the "parse" argument is set to true. This means that the response will be parsed by the library.
   */
  const response = await yt.actions.execute(continuation_node.endpoint.metadata.api_url, {
    ...continuation_node.endpoint.payload,
    parse: true
  });

  // Unwrap the response. Incremental continuation data is almost always inside special properties like this.
  const contents = response.on_response_received_endpoints.first().contents;

  // Print the response.
  console.log(contents);
  
  // Print the next continuation node (in case you can't/don't want to utilize the utility methods, the continuation is always at the bottom of the list):
  console.log(contents.firstOfType(YTNodes.ContinuationItem));
})();
```

The same concept can be applied to the many other feeds YouTube provides.
Let me know if this does what you need.

**Comment by Michal-Szczepaniak** [comment]
By your description it's exactly what i needed but let me test the code and get back to you! Thank a ton for help!

**Comment by Michal-Szczepaniak** [comment]
@LuanRT I was trying to apply your example to channel videos but i have some trouble. on_response_received_endpoints doesn't seem to exist in videos page

```
const channel = await youtube.getChannel(channelId)
const videos = await channel.getVideos();
let continuationData = videos.page.on_response_received_endpoints
```
continuationData is undefined

**Reply by LuanRT** [reply]
YouTube only sends `on_response_received_endpoints` for incremental continuations (when only a certain part of the page is updated). For pages such as channel tabs you are getting a partially full response, try printing the the page prop.

So in this case, you do not have to access the page prop though. It is much easier to find the continuation node in the `current_tab` prop (YouTube returns all the other tabs in the response too, so you'd have to find the selected one manually with the `page` prop).

Note:
For the channel page, I reckon you could also simply do `videos.memo.getType(YTNodes.ContinuationItem)`.

**Reply by Michal-Szczepaniak** [reply]
Oh i think i found it, it's `videos.current_tab.content.contents.firstOfType(YTNodes.ContinuationItem)`

tho i'll try your suggestion

**Comment by Michal-Szczepaniak** [comment]
I think i got it, exactly what i wanted

```
const channelId = process.argv[2];
const continuation = JSON.parse(process.argv[3])

let videos = null
let continuationData = null
if (continuation.length) {
    const response = await yt.actions.execute(continuation.endpoint.metadata.api_url, {
        ...continuation.endpoint.payload,
        parse: true
    });

    videos = response.on_response_received_actions_memo.getType(YTNodes.Video)
    continuationData = response.on_response_received_actions_memo.getType(YTNodes.ContinuationItem)
} else {
    const channel = await yt.getChannel(channelId)
    const channelVideos = await channel.getVideos()
    continuationData = channelVideos.memo.getType(YTNodes.ContinuationItem).at(0)
    videos = channelVideos.memo.getType(YTNodes.Video)
}

const result = {
    items: videos,
    continuation: continuationData
}

console.log(JSON.stringify(result, null, 2))
```

I think I now can apply same logic to other things as well. Thanks a ton for help!
